src/stackFrame_ppc64.cpp

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2017 Andrei Pangin
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(__PPC64__) && (__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__)
+
+#include "stackFrame.h"
+
+uintptr_t& StackFrame::pc() {
+    return (uintptr_t&)_ucontext->uc_mcontext.regs->nip;
+}
+
+uintptr_t& StackFrame::sp() {
+    return (uintptr_t&)_ucontext->uc_mcontext.regs->gpr[1];
+}
+
+uintptr_t& StackFrame::fp() {
+    return *((uintptr_t*)_ucontext->uc_mcontext.regs->gpr[1]);
+}
+
+uintptr_t StackFrame::arg0() {
+    return (uintptr_t)_ucontext->uc_mcontext.regs->gpr[3];
+}
+
+uintptr_t StackFrame::arg1() {
+    return (uintptr_t)_ucontext->uc_mcontext.regs->gpr[4];
+}
+
+uintptr_t StackFrame::arg2() {
+    return (uintptr_t)_ucontext->uc_mcontext.regs->gpr[5];
+}
+
+uintptr_t StackFrame::arg3() {
+    return (uintptr_t)_ucontext->uc_mcontext.regs->gpr[6];
+}
+
+void StackFrame::ret() {
+    _ucontext->uc_mcontext.regs->nip = _ucontext->uc_mcontext.regs->link;
+}
+
+static inline bool inC1EpilogueCrit(uintptr_t pc) {
+    // C1 epilogue and critical section (posX)
+    //        3821**** add     r1,r1,xx
+    // pos3   xxxxxxxx
+    // pos2   1000e1eb ld      r31,16(r1)
+    // pos1   a603e87f mtlr    r31
+    //        xxxxxxxx
+    //        2000804e blr
+    if (*((unsigned int *)pc + 1) == 0xebe10010 && *((unsigned int *)pc + 2) == 0x7fe803a6 ||
+        *((unsigned int *)pc + 0) == 0xebe10010 && *((unsigned int *)pc + 1) == 0x7fe803a6 ||
+        *((unsigned int *)pc - 1) == 0xebe10010 && *((unsigned int *)pc + 0) == 0x7fe803a6) {
```

Comment:
Is accessing memory before current PC always safe?
I can't tell about ppc, but on x86 and ARM there were situations where the address [pc-1] was not mapped.

Reply:
Not always, but as long as PC is within the VM's code cache it is. However, it it's a good idea to simply check if (PC-1) is within the code cache to be really robust. I'll add this check.